File: ports/pyagentic/pyagentic/runtime.py
"""Runtime — binds the agent core to an execution substrate.

``Runtime`` is the engine seam (the Python mirror of the Java "Engine SPI"): give it
an event, it builds an ``AgentContext`` from the shared stores and runs the graph,
guaranteeing single-writer-per-conversation. ``LocalRuntime`` is the zero-dependency
in-process implementation (a lock per conversation id); the Faust/Ray/etc. adapters
provide their own ``Runtime`` where the engine supplies the per-key ordering.
"""

from __future__ import annotations

from threading import Lock
from typing import Dict, Optional, Protocol, Union

from .core import Agent, AgentContext, Event, RoutedGraph, TurnResult
from .memory import ConversationStore, InMemoryConversationStore, InMemoryKeyedStateStore, KeyedStateStore
from .retrieval import TwoTierRetriever
from .tools import ToolRegistry

Handler = Union[Agent, RoutedGraph]


class Runtime(Protocol):
    def submit(self, event: Event) -> TurnResult: ...


class LocalRuntime:
    """In-process runtime: shared stores + a per-conversation lock so each
    conversation is processed by a single writer at a time (the local stand-in for
    Flink's keyBy / a Kafka partition / a Ray actor)."""

    def __init__(
        self,
        handler: Handler,
        store: Optional[ConversationStore] = None,
        state: Optional[KeyedStateStore] = None,
        tools: Optional[ToolRegistry] = None,
        retriever: Optional[TwoTierRetriever] = None,
    ) -> None:
        self.handler = handler
        self.store = store or InMemoryConversationStore()
        self.state = state or InMemoryKeyedStateStore()
        self.tools = tools or ToolRegistry()
        self.retriever = retriever
        self._locks: Dict[str, Lock] = {}
        self._locks_guard = Lock()

    def _lock_for(self, cid: str) -> Lock:
        with self._locks_guard:
            lk = self._locks.get(cid)
            if lk is None:
                lk = Lock()
                self._locks[cid] = lk
            return lk

    def submit(self, event: Event) -> TurnResult:
        ctx = AgentContext(
            conversation_id=event.conversation_id,
            user_id=event.user_id,
            store=self.store,
            state=self.state,
            tools=self.tools,
            retriever=self.retriever,
        )
        with self._lock_for(event.conversation_id):
            if isinstance(self.handler, RoutedGraph):
                return self.handler.handle(event, ctx)
            return self.handler.turn(event, ctx)

File: ports/experimental/go/core/runtime.go
package core

import "sync"

// LocalRuntime binds the agent core to the in-process substrate: shared stores + a
// per-conversation lock so each conversation is processed by a single writer at a time
// (the local stand-in for Flink's keyBy / a Kafka partition / a NATS subject). It is the
// zero-dependency Runtime; the NATS/Temporal adapters provide their own.
type LocalRuntime struct {
	graph     *RoutedGraph
	store     ConversationStore
	state     KeyedStateStore
	tools     *ToolRegistry
	retriever *TwoTierRetriever

	mu    sync.Mutex
	locks map[string]*sync.Mutex
}

// NewLocalRuntime builds a runtime. Pass nil for store/state to use in-memory defaults.
func NewLocalRuntime(graph *RoutedGraph, store ConversationStore, state KeyedStateStore,
	tools *ToolRegistry, retriever *TwoTierRetriever) *LocalRuntime {
	if store == nil {
		store = NewInMemoryConversationStore(0)
	}
	if state == nil {
		state = NewInMemoryKeyedStateStore()
	}
	if tools == nil {
		tools = NewToolRegistry()
	}
	return &LocalRuntime{
		graph: graph, store: store, state: state, tools: tools, retriever: retriever,
		locks: map[string]*sync.Mutex{},
	}
}

// NewBankingRuntime is a convenience constructor wiring the shared banking essence.
func NewBankingRuntime() *LocalRuntime {
	return NewLocalRuntime(BuildBankingGraph(), nil, nil, DefaultBankingTools(), BankingRetriever())
}

// Store exposes the conversation store (for transcript inspection by gateways).
func (r *LocalRuntime) Store() ConversationStore { return r.store }

func (r *LocalRuntime) lockFor(cid string) *sync.Mutex {
	r.mu.Lock()
	defer r.mu.Unlock()
	lk := r.locks[cid]
	if lk == nil {
		lk = &sync.Mutex{}
		r.locks[cid] = lk
	}
	return lk
}

// Submit runs one turn for an event, single-writer per conversation.
func (r *LocalRuntime) Submit(event Event) TurnResult {
	lk := r.lockFor(event.ConversationID)
	lk.Lock()
	defer lk.Unlock()
	ctx := &AgentContext{
		ConversationID: event.ConversationID,
		UserID:         event.UserID,
		Store:          r.store,
		State:          r.state,
		Tools:          r.tools,
		Retriever:      r.retriever,
	}
	return r.graph.Handle(event, ctx)
}
